app/code/Magento/Eav/Model/AttributeSetRepository.php

```diff
@@ -156,13 +170,19 @@ public function get($attributeSetId)
      */
     public function delete(AttributeSetInterface $attributeSet)
     {
+        // Get the affected product ids
+        $productIds = $this->productCollection
+            ->addFieldToFilter('attribute_set_id', $attributeSet->getAttributeSetId())->getAllIds();
+
         try {
             $this->attributeSetResource->delete($attributeSet);
+            $this->eventManager->dispatch('attribute_set_delete_after', ['products' => $productIds]);
```

Comment:
`Catalog` must not be referred in `Eav` module. New events must not be introduced when it's enough extension points to create a plugin-interceptor.

Reply:
You're right! I'll get on this later today 👍

Reply:
I don't think that the removal of an attribute set calls any delete function on a per product level, at least as far as I can tell.

Reply:
Products are dropped due to foreign keys? Maybe similar constraints should be added for product URL Rewrites?

Reply:
That isn't possible since there are other url rewrites in the table that relate to different entity types.

Reply:
You are right https://stackoverflow.com/a/28322144/8640867 :) That's our case, needs to be handled on application side, please implement as an interceptor attached to attribute set entity removal.